tokio/src/task/local.rs

```diff
@@ -416,88 +385,52 @@ impl Scheduler {
-        } else {
-            self.next_local_task().or_else(|| self.next_remote_task())
-        }
-    }
-
-    fn next_local_task(&self) -> Option<Task<Self>> {
-        unsafe { (*self.local_queue.get()).pop_front() }
-    }
-
-    fn next_remote_task(&self) -> Option<Task<Self>> {
-        // there is no semantic information in the `PoisonError`, and it
-        // doesn't implement `Debug`, but clippy thinks that it's bad to
-        // match all errors here...
-        #[allow(clippy::match_wild_err_arm)]
-        let mut lock = match self.remote_queue.lock() {
-            // If the lock is poisoned, but the thread is already panicking,
-            // avoid a double panic. This is necessary since `next_task` (which
-            // calls `next_remote_task`) can be called in the `Drop` impl.
-            Err(_) if std::thread::panicking() => return None,
-            Err(_) => panic!("mutex poisoned"),
-            Ok(lock) => lock,
-        };
-        lock.pop_front()
-    }
-
     fn tick(&self) {
         assert!(self.is_current());
         for _ in 0..MAX_TASKS_PER_TICK {
             let tick = self.tick.get().wrapping_add(1);
             self.tick.set(tick);
-            let task = match self.next_task(tick) {
+            let task = match unsafe { self.queues.next_task(tick) } {
                 Some(task) => task,
                 None => return,
             };
 
             if let Some(task) = task.run(&mut || Some(self.into())) {
                 unsafe {
                     // we are on the local thread, so this is okay.
-                    self.schedule_local(task);
+                    self.queues.push_local(task);
                 }
             }
         }
     }
-
-    fn drain_pending_drop(&self) {
-        for task in self.pending_drop.drain() {
-            unsafe {
-                (*self.tasks.get()).remove(&task);
-            }
-            drop(task);
-        }
-    }
-}
-
-impl fmt::Debug for Scheduler {
-    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
-        fmt.debug_struct("Scheduler { .. }").finish()
-    }
 }
 
 impl Drop for Scheduler {
     fn drop(&mut self) {
+        // Close the remote queue
+        self.queues.close_remote();
+
         // Drain all local tasks
-        while let Some(task) = self.next_local_task() {
+        while let Some(task) = unsafe { self.queues.next_local_task() } {
             task.shutdown();
         }
 
         // Release owned tasks
         unsafe {
-            (*self.tasks.get()).shutdown();
+            self.queues.shutdown();
         }
 
-        self.drain_pending_drop();
+        unsafe {
+            self.queues.drain_pending_drop();
+        }
 
         // Wait until all tasks have been released.
         // XXX: this is a busy loop, but we don't really have any way to park
         // the thread here?
-        while unsafe { !(*self.tasks.get()).is_empty() } {
-            self.drain_pending_drop();
+        unsafe {
```

Comment:
Hmm, what `Unpark` do the tasks use though? Could we not `Park` on that? Or do we not have one?

Reply:
The local scheduler has an `AtomicWaker`
https://github.com/tokio-rs/tokio/blob/7d78ba66863e737fbb7009866c35e638e9251774/tokio/src/task/local.rs#L90-L92
that's set from the context every time the `LocalFuture` is polled:
https://github.com/tokio-rs/tokio/blob/7d78ba66863e737fbb7009866c35e638e9251774/tokio/src/task/local.rs#L306

I don't think it's okay to use that while we're dropping, though?

Reply:
Not sure -- you'd have to check with @carllerche